src/lib/gadgets/range-check.ts

```diff
@@ -1,10 +1,56 @@
+import { Snarky } from '../../snarky.js';
+import { Fp } from '../../bindings/crypto/finite_field.js';
+import { Field as FieldProvable } from '../../provable/field-bigint.js';
 import { Field } from '../field.js';
 import { Gates } from '../gates.js';
-import { bitSlice, exists, toVar, toVars } from './common.js';
-
-export { rangeCheck64, multiRangeCheck, compactMultiRangeCheck };
+import { assert, bitSlice, exists, toVar, toVars } from './common.js';
+
+export {
+  rangeCheck64,
+  rangeCheck32,
+  multiRangeCheck,
+  compactMultiRangeCheck,
+  rangeCheckHelper,
+};
 export { l, l2, l3, lMask, l2Mask };
 
+/**
+ * Asserts that x is in the range [0, 2^32)
+ */
+function rangeCheck32(x: Field) {
+  if (x.isConstant()) {
+    if (x.toBigInt() >= 1n << 32n) {
+      throw Error(`rangeCheck32: expected field to fit in 32 bits, got ${x}`);
+    }
+    return;
+  }
+
+  let actual = rangeCheckHelper(32, x);
+  actual.assertEquals(x);
+}
+
+/**
+ * Create a new {@link Field} element from the first `length` bits of this {@link Field} element.
+ */
+function rangeCheckHelper(length: number, x: Field) {
```

Comment:
what about calling this `Gadgets.truncate` and putting into the common gadget section?

Reply:
again, `truncate()` suggests that the truncation is provable. But it's just the witness that happens to be the truncated version of the input -- no constraints enforce this. The only thing that constraints enforce is that the return value of `rangeCheckHelper` fits within 16*n bits. So, it's really just a "range check helper" and not truncation gadget.

My proposal: Create a gadget `rangeCheckMultiple16(x: Field, nBits: <multiple of 16>)` which uses `rangeCheckHelper(x).assertEquals(x)` internally, but remove `rangeCheckHelper` from the public API.

Also, would be nice to implement that gadget in TS on top of the raw EC scalar gate 👀

Reply:
EDIT: see https://github.com/o1-labs/o1js/pull/1259#discussion_r1407701112